motd change request button

benarmston opened the pull request:
This PR adds a button on the cluster's overview in the MOTD section to display the case form ready populated for creating a "MOTD change request".  

This works by adding `data-selected-category` and `data-seleted-issue` to the `new-case-form` DOM element which are passed into the elm app along with the other flags.  The elm `init` function then returns a list of `Cmd Msg`s to set the desired category and issue.

Whist this works, it would be better to have `decodeInitialModel` set the desired category and issue from `flags`.  Doing that would avoid a couple of trips through the elm runtime.  Unfortunately, how to do that is not immediately obvious to me.

Still to do: ~tests~

benarmston commented:
@bobwhitelock whilst writing the tests for this I discovered that some of the issue factories do not have an associated category.  Could you clarify when issues should have categories please?

benarmston commented on spec/features/case/new_spec.rb:
```diff
@@ -88,6 +88,7 @@
 
       visit new_case_path(as: user)
       fill_in 'New MOTD', with: motd
+      sleep 20
```
We don't.  I added it to better understand how the fixtures were working for that test. I thought I'd removed it before committing.  I'll do that now.

benarmston commented on app/controllers/cases_controller.rb:
```diff
@@ -35,6 +35,17 @@ def new
+                      category: tool.issue.category_id
+                    }
+                else
+                  @selected = {}
```
Good spot.

bobwhitelock commented:
Not sure if you've still got anything you want to add to this @benarmston, but if not happy for this to be merged as it is :)

benarmston commented:
I've decided implementing your 3rd suggested fix will be a good way of familiarising myself with elm.  If that goes well I'll have other things for you to look at.  If not, I'll bail on that and merge.  I'll let you know either way.

bobwhitelock commented:
OK sounds good, good luck :slightly_smiling_face:

benarmston commented:
@bobwhitelock I've now implemented your suggested solutions 2 and 3 above.  Doing so has given me a much better understanding of Elm and of the case form.  If you have any comments on the case form changes, I'd love to hear them.

As I had already implemented support for selecting a specific service, category, issue and tier, I have left in support for that.  Though I'm not sure that doing so is such a good idea.

The code adds a some complexity, that isn't needed to support selecting a tier 1 tool.  As the card I was working on only required support for selecting a tool.  It would perhaps be better to remove the code.  However, before I do so, I thought I'd mention it in case you know of any pending requirement for this.

The service, category, issue selecting code works mostly as expected, but does have some edge cases with issues that don't have a support type, such as "Suspected service issue".  If it worked without edge cases, I'd be more inclined to leave it.

bobwhitelock commented on app/javascript/packs/Service.elm:
```diff
@@ -67,3 +68,75 @@ setSelectedCategory : Category.Id -> Service -> Service
+                            Just ( service, Nothing, issue, tier )
+
+                        _ ->
+                            Nothing
```
In the different `case`s of `Maybe`s in this file you might be able to simplify things slightly using `Maybe.map` - would avoid you needing to unpack and then recreate the `Maybe`s yourself.

benarmston replied:
There is one point in that file where I have

```
                findTierAddingIssue issue =
                    Maybe.map ((,) issue) (Issue.findTier predicate issue)

```

but I couldn't find a general way of going from an n-tuple to an n+1-tuple.  Unless you know of the function that I need, I'm going to leave this as it is.

bobwhitelock commented on app/javascript/packs/State/Update.elm:
```diff
@@ -158,6 +161,48 @@ handleChangeTierField state index value =
+            state
+                |> .clusters
+                |> SelectList.selected
+                |> .services
```
This could be slightly simplified to `State.selectedCluster state |> .services` I believe.

benarmston replied:
Excellent. Thank you.

bobwhitelock commented on app/javascript/packs/State/Update.elm:
```diff
@@ -158,6 +161,48 @@ handleChangeTierField state index value =
+                ! []
+
+        Nothing ->
+            state ! []
```
Could use `Maybe.map` here I think so we only need to consider the stuff inside the `Maybe`.

benarmston replied:
Getting the types to line up with this is going to take a bit too much effort. So I'm going to leave it as it is.

bobwhitelock commented on app/javascript/packs/State/Update.elm:
```diff
@@ -158,6 +161,48 @@ handleChangeTierField state index value =
+                    state.clusters
+                        |> flip Cluster.setSelectedService service.id
+                        |> maybeSetCategory mcategory
+                        |> flip Cluster.setSelectedIssue (issue |> Issue.data |> .id)
```
A minor tweak, but could be worth pulling out an `Issue.id` method for extracting the `id` for symmetry with the other methods for unpacking Issue `data` (like `Issue.name` etc.)

benarmston replied:
Good idea.

bobwhitelock commented on app/javascript/packs/Service.elm:
```diff
@@ -67,3 +68,75 @@ setSelectedCategory : Category.Id -> Service -> Service
+    services
+        |> SelectList.map findServiceCategoryIssueAndTier
+        |> SelectList.Extra.find Maybe.Extra.isJust
+        |> Maybe.Extra.join
```
We do this same pattern in 3 places (`SelectList.map ... |> SelectList.Extra.find Maybe.Extra.isJust |> Maybe.Extra.join`), so could make things slightly simpler to extract a shared function for this.

benarmston replied:
Good idea.  We now have `SelectList.Extra.findValueBy`.  Not sure that's the best name, but it'll do for now.

bobwhitelock commented on spec/factories/tier.rb:
```diff
@@ -13,7 +13,8 @@
       factory :tier_with_tool do
-        tool :motd
+        association :issue, factory: :issue_with_category
+        tool 'motd'
```
Not that I mind the change just curious - was something breaking with using a symbol here? Will avoid using symbols in similar cases in future if so :slightly_smiling_face:

benarmston replied:
One of my tests was matching against that value.  I had `"motd" == :motd` causing the test to fail.  This change made the test work.  But on further thought I think the test would be more robust if I changed the test to call `.to_s`.  I'll make that change.

benarmston replied:
Or perhaps it was an entirely unnecessary change ;-)
